src/analysis/processing/qgsalgorithmsumlinelength.cpp

```diff
@@ -0,0 +1,214 @@
+}
+
+QString QgsSumLineLengthAlgorithm::inputParameterDescription() const
+{
+  return QObject::tr( "Polygons" );
+}
+
+QString QgsSumLineLengthAlgorithm::outputName() const
+{
+  return QObject::tr( "Line length" );
+}
+
+void QgsSumLineLengthAlgorithm::initParameters( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "LINES" ),
+                QObject::tr( "Lines" ), QList< int > () << QgsProcessing::TypeVectorLine ) );
+  addParameter( new QgsProcessingParameterString( QStringLiteral( "LEN_FIELD" ),
+                QObject::tr( "Lines length field name" ), QStringLiteral( "LENGTH" ) ) );
+  addParameter( new QgsProcessingParameterString( QStringLiteral( "COUNT_FIELD" ),
+                QObject::tr( "Lines count field name" ), QStringLiteral( "COUNT" ) ) );
+}
+
+bool QgsSumLineLengthAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  mLengthFieldName = parameterAsString( parameters, QStringLiteral( "LEN_FIELD" ), context );
+  mCountFieldName = parameterAsString( parameters, QStringLiteral( "COUNT_FIELD" ), context );
+
+  mLinesSource.reset( parameterAsSource( parameters, QStringLiteral( "LINES" ), context ) );
+  if ( !mLinesSource )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "LINES" ) ) );
+
+  if ( mLinesSource->hasSpatialIndex() == QgsFeatureSource::SpatialIndexNotPresent )
+    feedback->reportError( QObject::tr( "No spatial index exists for lines layer, performance will be severely degraded" ) );
+
+  return true;
+}
+
+QgsFields QgsSumLineLengthAlgorithm::outputFields( const QgsFields &inputFields ) const
+{
+  QgsFields outFields = inputFields;
+  mLengthFieldIndex = inputFields.lookupField( mLengthFieldName );
+  if ( mLengthFieldIndex < 0 )
+    outFields.append( QgsField( mLengthFieldName, QVariant::Double ) );
+
+  mCountFieldIndex = inputFields.lookupField( mCountFieldName );
+  if ( mCountFieldIndex < 0 )
+    outFields.append( QgsField( mCountFieldName, QVariant::Double ) );
+
+  mFields = outFields;
+  return outFields;
+}
+
+QgsFeatureList QgsSumLineLengthAlgorithm::processFeature( const QgsFeature &feature, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  QgsFeature outputFeature = feature;
+  if ( !feature.hasGeometry() )
+  {
+    QgsAttributes attrs = feature.attributes();
+    if ( mLengthFieldIndex < 0 )
+      attrs.append( 0 );
+    else
+      attrs[mLengthFieldIndex] = 0;
+
+    if ( mCountFieldIndex < 0 )
+      attrs.append( 0 );
+    else
+      attrs[mCountFieldIndex] = 0;
+
+    outputFeature.setAttributes( attrs );
+    return QList< QgsFeature > () << outputFeature;
+  }
+  else
+  {
+    const QgsGeometry polyGeom = feature.geometry();
+    std::unique_ptr< QgsGeometryEngine > engine( QgsGeometry::createGeometryEngine( polyGeom.constGet() ) );
+    engine->prepareGeometry();
+
+    QgsDistanceArea da;
```

Comment:
I'd move this up to the prepare step, or use a unique_ptr member so that we only need to initialize once (it's not cheap to create these)

Reply:
Done